File: cb/db/repositories/settings_repo.py
from __future__ import annotations
"""Settings repository -- simple key/value store in SQLite."""

from pathlib import Path
from typing import Optional

from cb.db.connection import get_connection


def get_setting(key: str, db_path: Optional[Path] = None) -> Optional[str]:
    conn = get_connection(db_path)
    try:
        row = conn.execute(
            "SELECT value FROM settings WHERE key = ?", (key,)
        ).fetchone()
        return row["value"] if row else None
    finally:
        conn.close()


def set_setting(key: str, value: str, db_path: Optional[Path] = None) -> None:
    conn = get_connection(db_path)
    try:
        conn.execute(
            "INSERT OR REPLACE INTO settings (key, value) VALUES (?, ?)",
            (key, value),
        )
        conn.commit()
    finally:
        conn.close()


def delete_setting(key: str, db_path: Optional[Path] = None) -> None:
    conn = get_connection(db_path)
    try:
        conn.execute("DELETE FROM settings WHERE key = ?", (key,))
        conn.commit()
    finally:
        conn.close()

File: src/core/db/repositories/settings-repo.ts
/**
 * Settings repository — simple key/value store in SQLite.
 * Behavioural 1:1 port of legacy/cb/db/repositories/settings_repo.py
 */

import { getConnection, initDb } from "../connection";

interface SettingsRow {
  value: string;
}

/**
 * Return the value for a settings key, or null if not found.
 * Auto-initialises the DB if the settings table is missing.
 * Mirrors Python get_setting().
 */
export function getSetting(key: string, dbPath?: string): string | null {
  const db = getConnection(dbPath);
  try {
    const row = db.query<SettingsRow, [string]>(
      "SELECT value FROM settings WHERE key = ?"
    ).get(key);
    return row !== null ? row.value : null;
  } catch (e) {
    if (String(e).includes("no such table")) {
      initDb(dbPath);
      return getSetting(key, dbPath);
    }
    throw e;
  } finally {
    db.close();
  }
}

/**
 * Insert or replace a setting.
 * Mirrors Python set_setting().
 */
export function setSetting(key: string, value: string, dbPath?: string): void {
  const db = getConnection(dbPath);
  try {
    db.run(
      "INSERT OR REPLACE INTO settings (key, value) VALUES (?, ?)",
      [key, value]
    );
  } finally {
    db.close();
  }
}

/**
 * Delete a setting by key.
 * Mirrors Python delete_setting().
 */
export function deleteSetting(key: string, dbPath?: string): void {
  const db = getConnection(dbPath);
  try {
    db.run("DELETE FROM settings WHERE key = ?", [key]);
  } finally {
    db.close();
  }
}
